tests/property/ic_api/canister_version/test/tests.ts

```diff
@@ -0,0 +1,129 @@
+import {
+    defaultPropTestParams,
+    expect,
+    getCanisterActor,
+    it,
+    Test
+} from 'azle/test';
+import { execSync } from 'child_process';
+import fc from 'fast-check';
+
+import { _SERVICE as Actor } from './dfx_generated/canister/canister.did';
+
+export function getTests(): Test {
+    return () => {
+        it('gets the canister version from the canister', async () => {
+            await fc.assert(
+                fc.asyncProperty(
+                    fc.nat({
+                        max: 10
+                    }),
+                    async (timesToDeploy) => {
+                        // Stop and delete the canister before every run
+                        execSync(`dfx canister stop canister || true`, {
+                            stdio: 'inherit'
+                        });
+
+                        execSync(`dfx canister delete canister || true`, {
+                            stdio: 'inherit'
+                        });
+
+                        execSync(`dfx deploy canister`, {
+                            stdio: 'inherit'
+                        });
+
+                        const startingVer = 1n; // Canister starts at 0 when it's created but the deploy in pretest should increment it to be at 1 by the time it gets here
+
+                        const actor = await getCanisterActor<Actor>('canister');
+
+                        await checkInitVersion(actor, startingVer);
+                        await checkUpgradeVersion(actor, []);
+                        await checkInspectVersion(actor, startingVer);
+                        await checkQueryAndUpdateVersion(
+                            actor,
+                            startingVer + 1n
+                        );
+
+                        const versionsAtDeploy = Array.from(
+                            { length: timesToDeploy },
+                            (_, index) => BigInt(index * 2) + startingVer
+                        );
+
+                        for (const version of versionsAtDeploy) {
+                            execSync(`dfx deploy canister --upgrade-unchanged`);
+
+                            await checkUpgradeVersion(actor, [version]);
+                            await checkQueryAndUpdateVersion(actor, version);
+                        }
+                    }
+                ),
+                defaultPropTestParams
+            );
+        });
+    };
+}
+
+async function checkInitVersion(
+    actor: Actor,
+    expectedVer: bigint
+): Promise<void> {
+    const initCanisterVersion = await actor.getInitCanisterVersion();
+
+    expect(initCanisterVersion).toHaveLength(1);
+    expect(initCanisterVersion[0]).toBeGreaterThanOrEqual(expectedVer);
```

Comment:
Shouldn't this be a strict equality check?

Reply:
no, unfortunately the ic gives us no guarantees that the canister version won't just arbitrarily increment for no reason at all... Or at least the way I read it we know that all of the events that cause it to increment will increment it, but also it might just increment for fun. So the best we can do is say we expect it to be greater than or equal to our expected value